Add type annotations to generate.py

isrqh opened the pull request:
Extracting type annotations from gh-28769 to separate the type annotations from thread-safety changes.

This adds type annotations to functions in `sympy/ntheory/generate.py` without changing any logic.

Still working on completing all function signatures.

Related to gh-28806, gh-28769

<!-- BEGIN RELEASE NOTES -->
* ntheory
  * Added type annotations to functions in generate module.
<!-- END RELEASE NOTES -->

isrqh committed:
Add type annotations to generate.py (WIP)

Extracting type annotations from #28769.
Still working on adding all function signatures.

Related to #28806, #28769

isrqh committed:
Add type annotations to generate.py

Added type hints to 11 public functions in ntheory/generate.py:
- prime, primepi, _primepi, nextprime, prevprime, primerange,
randprime, primorial, composite, compositepi, and cycle_length.

These are extracted from gh-28769 to simplify review.

Issue: https://github.com/sympy/sympy/issues/28806

oscarbenjamin commented:
> using `SupportsIndex` with `as_int()` conversions - should I update this PR to use that approach instead?

Yes.

> Try to add the `@overload` signatures here?

Yes, it should use `@overload` like in the other PR.

isrqh committed:
Use SupportIndex and add cycle_length overloads
- Replaced int with SupportsIndex for all numeric parameters
- Add as_int() conversions throughout
- Add @overload to cycle_length
- Rename variables for clarity (n-> n_int, a-> a_int, b-> b_int)
- Add required type imports (Callable, overload, SupportsIndex, Literal)

No changes to core logic or algorithms. All tests pass and mypy reports no errors.

isrqh committed:
Add SupportsIndex type annotations per review feedback

- Updated function signatures to accept SupportsIndex
- Added wrapper functions for type conversion

All tests passing.

isrqh committed:
Remove trailing whitespace

M-Israr-Ul-Haq commented:
Applied the wrapper pattern for `nextprime ` and `primerange` as suggested. Other functions just got `SupportsIndex `in their signatures with conversions at the beginning.

oscarbenjamin commented on sympy/ntheory/generate.py:
```diff
@@ -836,6 +840,12 @@ def primerange(a, b=None):
+def _primerange(a: int, b: int | None = None) -> Iterator[int]:
+    """ Internal implementation of primerange. """
     if b is None:
         a, b = 2, a
```
It would be better to handle that `None` case in `primerange` so that the arguments to `_primerange` are simple.

isrqh committed:
Move None handling to primerange function

_primerange now always receives int arguments.

oscarbenjamin commented:
> I extracted the type annotations from your [gh-28769](https://github.com/sympy/sympy/pull/28769) as requested in [gh-28806](https://github.com/sympy/sympy/issues/28806).

The other PR also adds annotations to everything in the Sieve class. It would be good to include those here as well.

oscarbenjamin commented:
Don't change the methods or add any new methods. Just add the type annotations to the things that the Sieve class has in the master branch (including the attributes like `_tlist`).

M-Israr-Ul-Haq commented:
@oscarbenjamin Would it make sense to type the argument as `Expr `in `_as_int_ceiling`, or should i leave it untyped to match the current behavior? 
```
def _as_int_ceiling(a) -> int:
```
I think leaving it untyped is reasonable.

isrqh committed:
Add type annotations to Sieve class

- Annotated class attributes and all method signatures
- Added @overload for __getitem__ (int / slice)
- Types only, no logic or behavior changes
